FLT/DedekindDomain/AdicValuation.lean

```diff
@@ -300,4 +302,69 @@ theorem closureAlgebraMapIntegers_eq_prodIntegers {ι : Type*}
       exact Set.mem_range_self a
     . exact hts fun w hw ↦ hg w <| ha w hw
 
-end IsDedekindDomain
+lemma adicCompletion.eq_mul_nonZeroDivisor_inv_adicCompletionIntegers (v : HeightOneSpectrum A)
+    (x : v.adicCompletion K) :
+    ∃a ∈ nonZeroDivisors A, ∃b ∈ v.adicCompletionIntegers K, x = (algebraMap A K a)⁻¹ • b := by
+  obtain ⟨a, hz, ha⟩ :=
+    adicCompletion.mul_nonZeroDivisor_mem_adicCompletionIntegers v x
+  use a, hz, (algebraMap A K a) • x
+  constructor
+  . rwa [Algebra.smul_def, ← IsScalarTower.algebraMap_apply, mul_comm]
+  . rw [smul_smul, inv_mul_cancel₀, one_smul]
+    exact IsLocalization.to_map_ne_zero_of_mem_nonZeroDivisors K (fun _ ↦ id) hz
+
+lemma adicCompletion.eq_mul_pi_adicCompletionIntegers {ι : Type*} [Fintype ι]
+    (valuation : ι → HeightOneSpectrum A) (x : (i : ι) → (valuation i).adicCompletion K) :
+      ∃k : K, ∃y ∈ Set.pi Set.univ (fun (i : ι) ↦ ((valuation i).adicCompletionIntegers K).carrier),
+      x = k • y := by
+  classical
+  choose f hf using fun (i : ι) =>
+    eq_mul_nonZeroDivisor_inv_adicCompletionIntegers K (valuation i) (x i)
+  use (algebraMap A K (∏ i : ι, f i))⁻¹, (algebraMap A K (∏ i : ι, f i)) • x
+  have hz : ∀ (i : ι), (algebraMap A K) (f i) ≠ 0 := fun i =>
+    IsLocalization.to_map_ne_zero_of_mem_nonZeroDivisors K (fun _ ↦ id) (hf i).left
+  constructor
+  . rintro i -
+    obtain ⟨b, hb, hx⟩ := (hf i).right
+    beta_reduce
+    rw [Pi.smul_apply, algebraMap_smul, Subsemiring.coe_carrier_toSubmonoid,
+        Subring.coe_toSubsemiring, SetLike.mem_coe, ValuationSubring.mem_toSubring, hx,
+        ← Finset.prod_erase_mul _ f (Finset.mem_univ i), mul_smul,
+        ← IsScalarTower.smul_assoc (f i), Algebra.smul_def (f i), mul_inv_cancel₀ (hz i), one_smul,
+        Algebra.smul_def]
+    apply mul_mem (coe_mem_adicCompletionIntegers _ _) hb
+  . rw [smul_smul, inv_mul_cancel₀, one_smul]
+    simp [Finset.prod_ne_zero_iff, hz]
+
+/-- If `s` is finite then `K` in dense in `∏_{v ∈ s} K_v`. -/
+theorem denseRange_of_prodAlgebraMap {ι : Type*} [Fintype ι]
+    {valuation : ι → HeightOneSpectrum A} (injective : Function.Injective valuation) :
+    DenseRange (algebraMap K ((i : ι) → (valuation i).adicCompletion K)) := by
+  rw [denseRange_iff_closure_range]
+  let S := Set.range (algebraMap K ((i : ι) → (valuation i).adicCompletion K))
+  let I := Set.pi Set.univ (fun (i : ι) ↦ ((valuation i).adicCompletionIntegers K).carrier)
+  -- We've already shown that the closure of `A` is `∏_{v ∈ s} 𝒪_v`, so
+  -- the closure of `K` at least contains this set.
+  have hint : I ⊆ closure S := by
+    unfold I
```

Comment:
I always think of `unfold` as a bit of a code smell. You can avoid it here with `set I := Set.Pi ... carrier) with hI` and then you can `rw [hI]`. But I'm not demainding this, I'm really just letting you know about the idiom.